src/bloqade/builder/assign.py

```diff
@@ -8,14 +8,26 @@
 from decimal import Decimal
 
 
-def cast_param(value: Union[Real, List[Real]]):
-    if isinstance(value, list):
-        return list(map(Decimal, map(str, value)))
-
+def cast_scalar_param(value: Union[Real, Decimal]) -> Decimal:
     if isinstance(value, (Real, Decimal)):
         return Decimal(str(value))
 
-    raise ValueError("value must be a real number or a list of real numbers")
+    raise TypeError(
+        "assign parameter value must be a real number, " f"found type: {type(value)}"
+    )
+
+
+def cast_batch_param(value: List[Real]) -> List[Decimal]:
+    if isinstance(value, (list, tuple)):
+        return list(map(cast_scalar_param, value))
+
+    if isinstance(value, np.ndarray):
+        return list(map(cast_scalar_param, value.tolist()))
+
+    raise TypeError(
+        "batch_assign parameter value must be a list of real numbers, "
```

Comment:
Would it be worth being more explicit here? e.g. "batch_assign parameter value must be a list/**tuple** of real numbers,..."?